### PR TITLE
routing: send to route max hops check

### PR DESCRIPTION
**Pull request by joostjager**
Depends on https://github.com/lightningnetwork/lightning-onion/pull/45

A follow up pr https://github.com/lightningnetwork/lnd/pull/3841 will also limit path finding to the maximum payload size of 1300 bytes.

Fixes https://github.com/lightningnetwork/lnd/issues/3712

### CODE REVIEW

**Comment by Roasbeef** [comment]
Let's also update this to pin to the latest version of `lightning-onion` created here: https://github.com/lightningnetwork/lightning-onion/pull/45

**Comment by joostjager** [comment]
Updated to f34e9dc1651d174d79cc7a9965ac26844eb4ec98

File: routing/payment_lifecycle.go [hunk 1]
```diff
@@ -279,8 +281,28 @@ func (p *paymentLifecycle) createNewPaymentAttempt() (lnwire.ShortChannelID,
...
+		err == sphinx.ErrMaxRoutingInfoSizeExceeded {
+
+		log.Debugf("Invalid route provided for "+
+			"payment %x: %v", p.payment.PaymentHash,
+			err)
+
+		controlErr := p.router.cfg.Control.Fail(
+			p.payment.PaymentHash,
+			channeldb.FailureReasonError,
+		)
```
Comment (guggero) [inline routing/payment_lifecycle.go]: nit: those two statements are both wrapped a bit too early and could be more compact if some line breaks are removed.

File: routing/route/route.go [hunk 2]
```diff
@@ -267,6 +271,11 @@ func NewRouteFromHops(amtToSend lnwire.MilliSatoshi, timeLock uint32,
 func (r *Route) ToSphinxPath() (*sphinx.PaymentPath, error) {
 	var path sphinx.PaymentPath
 
+	// Check maximum route length.
+	if len(r.Hops) > sphinx.NumMaxHops {
```
Comment (guggero) [inline routing/route/route.go]: Could we move the `len(rt.Hops) == 0` check from `generateSphinxPacket` to here as well? It returns an error from the `route` package anyway.

Comment (joostjager) [reply routing/route/route.go]: Nice suggestion, done

**Pull request merged**